gym/vector/utils/spaces.py

```diff
@@ -87,3 +88,92 @@ def batch_space_dict(space, n=1):
 
 def batch_space_custom(space, n=1):
     return Tuple(tuple(space for _ in range(n)))
+
+
+def iterate(items, space):
+    """Iterate over the elements of a (batched) space.
+
+    Parameters
+    ----------
+    items : samples of `space`
+        Items to be iterated over.
+
+    space : `gym.spaces.Space` instance
+        Space to which `items` belong to.
+
+    Returns
+    -------
+    iterator : `Iterable` instance
+        Iterator over the elements in `items`.
+
+    Example
+    -------
+    >>> from gym.spaces import Box, Dict
+    >>> space = Dict({
+    ... 'position': Box(low=0, high=1, shape=(2, 3), dtype=np.float32),
+    ... 'velocity': Box(low=0, high=1, shape=(2, 2), dtype=np.float32)})
+    >>> items = space.sample()
+    >>> it = iterate(items, space)
+    >>> next(it)
+    {'position': array([-0.99644893, -0.08304597, -0.7238421 ], dtype=float32),
+    'velocity': array([0.35848552, 0.1533453 ], dtype=float32)}
+    >>> next(it)
+    {'position': array([-0.67958736, -0.49076623,  0.38661423], dtype=float32),
+    'velocity': array([0.7975036 , 0.93317133], dtype=float32)}
+    >>> next(it)
+    StopIteration
+    """
+    if isinstance(space, _BaseGymSpaces):
```

Comment:
Why not put the space as the first argument, and use a singledispatch callable here? 
This would let users customize how this function should behave with their custom spaces, which isn't possible as it is.

Reply:
You're also essentially doing the same thing as a singledispatch here, but worse! :P

Reply:
It is definitely a prime candidate for singledispatch!
I was waiting for #2093, but I'll get the ball rolling and switch this function to singledispatch already.